cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/ReplicaDistributionGoal.java

```diff
@@ -297,18 +284,22 @@ protected void rebalanceForBroker(Broker broker,
                                     Set<String> excludedTopics) {
     LOG.debug("Rebalancing broker {} [limits] lower: {} upper: {}.", broker.id(), _balanceLowerLimit, _balanceUpperLimit);
     int numReplicas = broker.replicas().size();
-    boolean requireLessReplicas = broker.isAlive() ? numReplicas > _balanceUpperLimit : numReplicas > 0;
-    boolean requireMoreReplicas = broker.isAlive() && numReplicas < _balanceLowerLimit;
+    int numOfflineReplicas = broker.currentOfflineReplicas().size();
+
+    // A broker with a bad disk may take new replicas and give away offline replicas.
+    boolean requireLessReplicas = numOfflineReplicas > 0 || (numReplicas > _balanceUpperLimit);
+    boolean requireMoreReplicas = broker.isAlive() && numReplicas - numOfflineReplicas < _balanceLowerLimit;
     if (broker.isAlive() && !requireMoreReplicas && !requireLessReplicas) {
-      // return if the broker is already under limit.
+      // return if the broker is already within the limit.
       return;
-    } else if (!clusterModel.newBrokers().isEmpty() && requireMoreReplicas && !broker.isNew()) {
-      // return if we have new brokers and the current broker is not a new broker but require more load.
+    } else if (!clusterModel.newBrokers().isEmpty() && requireMoreReplicas && !broker.isNew() && !requireLessReplicas) {
+      // return if we have new brokers and the current broker is not a new broker but requires: more replicas, but not
+      // less replicas -- i.e. hence, does not have offline replicas on it.
       return;
-    } else if (!clusterModel.deadBrokers().isEmpty() && requireLessReplicas && broker.isAlive()
-        && broker.immigrantReplicas().isEmpty()) {
+    } else if (!clusterModel.selfHealingEligibleReplicas().isEmpty() && requireLessReplicas
```

Comment:
If there are offline replicas, the first attempt tries to move only offline or immigrant replicas away from dead brokers or broken disks *while also tying to satisfy goal requirements*.
Whereas if `_fixOfflineReplicasOnly` is true, then the optimization will omit the replica balance limit to relocate all remaining replicas away from offline disks / brokers.
Overall, this approach provides a faster resolution in case of self-healing.

Reply:
Thanks for explanation. A more general question: for all the goals,  in self-healing mode, in first attempt, if there is no optimization can be done on offline and immigrant replica, and the goal's requirement has not be reached, no optimization on other replica will be applied,  is this assumption correct?

Reply:
For hard goals, all replicas are considered independent of whether they are self-healing eligible or not; hence, this assumption does not hold. For the majority of soft-goals this is the case, but there are also soft goals that do not even consider self-healing -- e.g. `LeaderBytesInDistributionGoal`).